README.md
# LDAP query generator

Writing LDAP queries is hard!
this is a tool to generate LDAP operations defined in [RFC 4511](https://tools.ietf.org/html/rfc4511)

## AS Easy AS

```ts
import { QueryGenerator } from "ldap-query-generator";

/** User Fields */
interface User {}

/** You can use it with or without generic type */
const qGen = new QueryGenerator<User>();
const { query } = qGen
  .select(["USNIntersite", "aCSPolicyName"])
  .where({ field: "mobile", action: "substrings", criteria: "404*999*" })
  .whereAnd({ field: "memberOf", action: "startWith", criteria: "admin" })
  .whereAnd({ field: "memberOf", action: "endWith", criteria: "office" })
  .whereAnd({ field: "badPwdCount", action: "lessOrEqual", criteria: "2" })
  .whereAnd({ field: "info", action: "approxMatch", criteria: "my-info" })
  .whereOr({ field: "mail", action: "present", criteria: "*@domain.com" })
  .whereOr({
    field: "homePostalAddress",
    action: "substrings",
    criteria: "Georgia",
  })
  .whereNot({
    field: "delivContLength",
    action: "greaterOrEqual",
    criteria: "6",
  })
  .whereNot({
    field: "middleName",
    action: "extensible",
    criteria: "joe",
    extensibleConfig: {
      dn: true,
      ignoreField: true,
      matchingRuleId: "1.2.840.113556.1.4.1941",
    },
  })
  .whereNot({
    field: "userAccountControl",
    action: "extensible",
    criteria: "2",
    extensibleConfig: {
      dn: false,
      ignoreField: false,
      matchingRuleId: "1.2.840.113556.1.4.803",
    },
  })
  .whereRaw("&(cn=3)(dn=*)")
  .whereRaw("phone=*11");

console.log(query.toString());
```

Output:

```
(&(mobile=404*999*)(&(memberOf=admin*))(&(memberOf=*office))(&(badPwdCount<=2))(&(info~=my-info))(|(mail=*))(|(homePostalAddress=Georgia))(!(delivContLength>=6))(!(:dn:1.2.840.113556.1.4.1941:=joe))(!(userAccountControl:1.2.840.113556.1.4.803:=2))(&(cn=3)(dn=*))(phone=*11))
```

### Note:

to generate interfaces from ldap schema, use [ldap-schema-ts-generator](https://www.npmjs.com/package/ldap-schema-ts-generator)

### Api Documentations

API documentation [API Website](https://saostad.github.io/ldap-query-generator/)

## TODO

- [ ] LDAP Search Filters [RFC4515](https://tools.ietf.org/html/rfc4515)
  - [x] where
  - [x] whereAnd
  - [x] whereOr
  - [x] whereNot
  - [x] select
  - [x] toString
  - [x] whereRaw
  - [ ] Absence of attribute (!(attribute=_)) , e.g. (!proxyAddresses=_)
  - [ ] Filter boolean attributes the consideration of the upper/ lower case will be crucial. The use of TRUE or FALSE is absolutely necessary for filtering such booleans.
  - [ ] Special characters: characters ( ) & | = ! > < ~ \* / \ play a special role for the declaration of LDAP filters.
  - [ ] Hex Numbers
  - [ ] Binary Values
  - [ ] Filtering for Bit Fields
  - [ ] Filtering with Ambiguous Name Resolution (ANR)
- [ ] LDAP Search Filter Validator
  - [ ] No quotation marks Comparative strings do NOT appear in quotation marks. A filter for the displayName 'Philipp Foeckeler' would read as follows: (displayName=Philipp Foeckeler)
  - [ ] correct parentheses
  - [ ] you can't use wildcards in LDAP filters for attributes containing LDAP distinguished names (attributes with DN-string syntax / ADSI attribute data type ADSTYPE_DN_STRING = 1). The same applies for ADS: Filters in which DN attributes are searched with wildcards do not work. the following filter won't work!
        (distinguishedName=\*,ou=Sydney,dc=cerrotorre,dc=org)

## Inspired By:

- [PHP LdapTools](http://www.phpldaptools.com/tutorials/Building-LDAP-Queries/)
- [knex.js](http://knexjs.org/)
- [ldap.js](http://ldapjs.org/filters.html)
- [Adldap2](https://github.com/Adldap2/Adldap2)

### Useful Resources

- [ldapexplorer](http://www.ldapexplorer.com/en/manual/109010000-ldap-filter-syntax.htm)


CHANGELOG.md
### Changelog

All notable changes to this project will be documented in this file. Dates are displayed in UTC.

#### [v0.8.7](https://github.com/saostad/ldap-query-generator/compare/v0.8.6...v0.8.7)

> 17 August 2022

- pkg: updated [`f861890`](https://github.com/saostad/ldap-query-generator/commit/f8618907ef2e11769d8ec2e4e8c8f93e63526451)
- doc: change log [`83ab660`](https://github.com/saostad/ldap-query-generator/commit/83ab66055b65a4f100d48616103478de9bb63fdf)

#### [v0.8.6](https://github.com/saostad/ldap-query-generator/compare/v0.8.5...v0.8.6)

> 16 August 2022

- pkg: updated [`d481ce5`](https://github.com/saostad/ldap-query-generator/commit/d481ce53f0668361e4d5bb2679be8eca55f7950a)
- doc: change log [`1bdabc6`](https://github.com/saostad/ldap-query-generator/commit/1bdabc6d8360cdc6f1115963fdad803d325a58af)

#### [v0.8.5](https://github.com/saostad/ldap-query-generator/compare/v0.8.4...v0.8.5)

> 16 August 2022

- pkg: updated [`a21516a`](https://github.com/saostad/ldap-query-generator/commit/a21516a5d40d86fdc0114e14738a8c4ec89a1a7c)
- doc: change log [`978f078`](https://github.com/saostad/ldap-query-generator/commit/978f0789663b56c8cf03dc029d8c3c3c2b668738)

#### [v0.8.4](https://github.com/saostad/ldap-query-generator/compare/v0.8.3...v0.8.4)

> 20 July 2022

- pkg: updated [`10f163a`](https://github.com/saostad/ldap-query-generator/commit/10f163af139b068ea9d858f9129942f0ebb9fb86)
- doc: change log [`96bb755`](https://github.com/saostad/ldap-query-generator/commit/96bb75504140bffc4f52495e56810cf05a4c4a10)

#### [v0.8.3](https://github.com/saostad/ldap-query-generator/compare/v0.8.2...v0.8.3)

> 2 March 2021

- pkg: updated [`6203eab`](https://github.com/saostad/ldap-query-generator/commit/6203eab71b5142da579a6f7fa9b0b872b088e4f9)
- doc: change log [`256ffdb`](https://github.com/saostad/ldap-query-generator/commit/256ffdb0c915b43a2b30c3d1a4d7fe44b9d033e2)

#### [v0.8.2](https://github.com/saostad/ldap-query-generator/compare/v0.8.1...v0.8.2)

> 28 December 2020

- fix: ExtensibleMatch bug fix [`e71c63b`](https://github.com/saostad/ldap-query-generator/commit/e71c63bdfcc9c5e33b5f7e980e650b09a606e036)
- package: updated [`3a4b2a1`](https://github.com/saostad/ldap-query-generator/commit/3a4b2a1e61f176258ea3cc16e5ba349035b6887f)
- package: updated [`3c992cd`](https://github.com/saostad/ldap-query-generator/commit/3c992cd813a425d4839fb84c411e1a053f5b3fad)

#### [v0.8.1](https://github.com/saostad/ldap-query-generator/compare/v0.8.0...v0.8.1)

> 7 June 2020

- package: updated [`884ab8e`](https://github.com/saostad/ldap-query-generator/commit/884ab8e1966c1273d8c15eda1d697c1c51254e03)
- fix: bug in generated string [`83e3862`](https://github.com/saostad/ldap-query-generator/commit/83e386231f41da6efa1945d5a22fcd488884fc00)
- doc: todo list updated [`a19bf21`](https://github.com/saostad/ldap-query-generator/commit/a19bf21381996612930651ec6845317fc1691f44)

#### v0.8.0

> 25 April 2020

- **Breaking change:** pub: 0.7.0 [`9ca2d79`](https://github.com/saostad/ldap-query-generator/commit/9ca2d79e11eb1be641a484f8b4f784e1266fde27)
- init commit [`414bd2f`](https://github.com/saostad/ldap-query-generator/commit/414bd2fe269651e3e740dc0cc475e4c5627eef1b)
- [pub] 0.1.0 [`32d6f31`](https://github.com/saostad/ldap-query-generator/commit/32d6f31950b813ccc1fd9d0b8d255b348520365a)


src/index.ts
import type { Logger } from "fast-node-logger";
import { toString } from "./helpers/toString";
import { CriteriaActions } from "./helpers/criteria";

export type Query<T = any> = {
  /** selected attributes to return */
  attributes?: SelectInput<T>[];
  /** one time only */
  where?: WhereInput;
  /** can be use multiple time */
  whereAnd: WhereInput[];
  /** can be use multiple time */
  whereOr: WhereInput[];
  /** can be use multiple time */
  whereRaw: string[];
  /** can be use multiple time */
  whereNot: WhereInput[];
  /** generate ldap filter */
  toString: () => string;
};

export type WhereInput<T = any> = {
  field: Extract<keyof T, string>;
  /**
   * - '*' retrieve all objects with attribute
   * - '!*' retrieve all objects do not have attribute
   */
  criteria: string;
  action: CriteriaActions;
  extensibleConfig?: {
    ignoreField?: boolean;
    dn?: boolean;
    matchingRuleId?: string;
  };
};

type GeneratorInput = {
  /** instance of logger compatible with pino */
  logger?: Logger;
};

type SelectInput<T = any> = Extract<keyof T, string> | "*";

/** query generator instance */
export class QueryGenerator<T = any> {
  public query!: Query<T>;
  private logger?: Logger;

  constructor(options?: GeneratorInput) {
    this.logger = options?.logger;

    this.query = {
      where: undefined,
      whereAnd: [],
      whereNot: [],
      whereRaw: [],
      whereOr: [],
      attributes: [],
      toString: () => toString(this.query),
    };
  }

  /** (cn=foo) */
  public where(input: WhereInput<T>) {
    this.logger?.trace(`where()`);
    if (this.query?.where) {
      throw new Error(
        `where can be one time use! use whereAnd, whereOr or whereNot instead`,
      );
    }
    this.query.where = input;
    return this;
  }

  /** whatever you decide will be wrap in () and add to the end of generated filter string.
   * @example &(field1=1234)(field2=*) result: (&(field1=1234)(field2=*))
   * @note it supposed you know what are you doing!
   */
  public whereRaw(input: string) {
    this.logger?.trace(`whereRaw()`);
    this.query.whereRaw?.push(input);
    return this;
  }

  /** (&(cn=foo)(sn=bar)) */
  public whereAnd(input: WhereInput<T>) {
    this.logger?.trace(`whereAnd()`);
    this.query.whereAnd?.push(input);
    return this;
  }

  /** (|(cn=foo)(sn=bar)) */
  public whereOr(input: WhereInput<T>) {
    this.logger?.trace(`whereOr()`);
    this.query.whereOr?.push(input);
    return this;
  }

  /** (!(cn=foo)) */
  public whereNot(input: WhereInput<T>) {
    this.logger?.trace(`whereNot()`);
    this.query.whereNot?.push(input);
    return this;
  }

  /** select return attributes
   * - put ["*"] for select all available attributes */
  public select(fields: Array<SelectInput<T>>) {
    this.logger?.trace(`select()`);
    this.query.attributes = this.query.attributes?.concat(fields);
    return this;
  }
}


src/helpers/criteria.ts
export type CriteriaActions =
  | "equal"
  | "approxMatch"
  | "substrings"
  | "greaterOrEqual"
  | "lessOrEqual"
  | "present"
  | "extensible"
  | "startWith"
  | "endWith";

type CriteriaActionsFnInput = {
  field: string;
  criteria: string;
};
/** equalityMatch */
export function equal({ field, criteria }: CriteriaActionsFnInput) {
  if (criteria.includes("*")) {
    throw new Error(
      `equal action can't accept criteria ${criteria}. you should use substring if you want to use * in criteria.`,
    );
  }
  return `${field}=${criteria}`;
}

/** (cn=jo*) */
export function startWith({ field, criteria }: CriteriaActionsFnInput) {
  return `${field}=${criteria}*`;
}

/** (cn=j*o) */
export function endWith({ field, criteria }: CriteriaActionsFnInput) {
  return `${field}=*${criteria}`;
}

/** @description The string representation of an ApproxMatch filter is constructed as follows:
 * An open parenthesis The attribute description (potentially including attribute options) A tilde character An equal sign The value to compare (aka the assertion value) A close parenthesis.
 * @example
 * (givenName~=John)
 * /% to match entries with givenName values of either John or Jon.
 * @note doesn't work with Microsoft Active Directory. Active Directory supports the approxMatch filter clause of [RFC2251] section 4.5.1. However, it is implemented identically to equalityMatch; for example, the filter is true if the values are equal. No approximation is performed. Filter clauses of the form "(X=Y)" and "(X~=Y)" can be freely substituted for each other.
 * @note Although it seems like a significant oversight or omission, the LDAP specifications do not make any provision for ApproxMatch matching Rules. A number of individual LDAP Server Implementations provide this capability anyway so that it may be possible to configure the approximate match behavior on a per-attribute basis, but the inconsistency of approximate matching capabilities between server implementations makes approximate matching something that is often avoided in LDAP-enabled applications.
 */
export function approxMatch({ field, criteria }: CriteriaActionsFnInput) {
  return `${field}~=${criteria}`;
}

/**
 *  "(cn=ab*def*mno*stu*yz)"
 * contains a subInitial component of "ab", subAny components of
 * "def", "mno", and "stu", and a subFinal component of "yz".
 */
export function substrings({ field, criteria }: CriteriaActionsFnInput) {
  return `${field}=${criteria}`;
}

export function greaterOrEqual({ field, criteria }: CriteriaActionsFnInput) {
  return `${field}>=${criteria}`;
}

export function lessOrEqual({ field, criteria }: CriteriaActionsFnInput) {
  return `${field}<=${criteria}`;
}

type PresentFnInput = {
  field: string;
};
/** field exist */
export function present({ field }: PresentFnInput) {
  return `${field}=*`;
}

type ExtensibleFnInput = {
  field?: string;
  dn?: boolean;
  matchingRuleId?: string;
  criteria: string;
};
/**
 * [More info](https://ldapwiki.com/wiki/ExtensibleMatch)
 * @example
 * (cn:1.2.3.4.5:=Fred Flintstone)
 * (sn:dn:2.4.6.8.10:=Barney Rubble)
 * (o:dn:=Ace Industry)
 * (:dn:2.4.6.8.10:=Dino)
 */
export function extensible({
  field,
  dn,
  matchingRuleId,
  criteria,
}: ExtensibleFnInput) {
  return `${field ?? ""}${dn ? ":dn" : ""}:${
    matchingRuleId ?? ""
  }:=${criteria}`;
}


src/helpers/toString.ts
import { Query, WhereInput } from "../index";
import {
  equal,
  substrings,
  present,
  lessOrEqual,
  greaterOrEqual,
  extensible,
  approxMatch,
  endWith,
  startWith,
} from "./criteria";
import { writeLog } from "fast-node-logger";

function whereClausesToString({
  field,
  criteria,
  action,
  extensibleConfig,
}: WhereInput) {
  writeLog(`whereClausesToString()`, { level: "trace" });
  if (action === "equal") {
    return equal({ field: field as string, criteria });
  }
  if (action === "startWith") {
    return startWith({ field: field as string, criteria });
  }
  if (action === "endWith") {
    return endWith({ field: field as string, criteria });
  }
  if (action === "approxMatch") {
    return approxMatch({ field: field as string, criteria });
  }
  if (action === "greaterOrEqual") {
    return greaterOrEqual({ field: field as string, criteria });
  }
  if (action === "lessOrEqual") {
    return lessOrEqual({ field: field as string, criteria });
  }
  if (action === "present") {
    return present({ field: field as string });
  }
  if (action === "substrings") {
    return substrings({ field: field as string, criteria });
  }
  if (action === "extensible") {
    return extensible({
      field: extensibleConfig?.ignoreField ? undefined : (field as string),
      criteria,
      dn: extensibleConfig?.dn,
      matchingRuleId: extensibleConfig?.matchingRuleId,
    });
  }
}

/** generate ldap representation of query */
export function toString(query: Query): string {
  writeLog(`toString()`, { level: "trace" });
  const result: string[] = [];

  if (query.where) {
    result.push(`(${whereClausesToString(query.where)})`);
  }

  if (query.whereAnd.length > 0) {
    result.push(
      `${query.whereAnd
        .map((el) => `(&(${whereClausesToString(el)}))`)
        .join("")}`,
    );
  }

  if (query.whereOr.length > 0) {
    result.push(
      `${query.whereOr
        .map((el) => `(|(${whereClausesToString(el)}))`)
        .join("")}`,
    );
  }

  if (query.whereNot.length > 0) {
    result.push(
      `${query.whereNot
        .map((el) => `(!(${whereClausesToString(el)}))`)
        .join("")}`,
    );
  }

  if (query.whereRaw.length > 0) {
    result.push(`${query.whereRaw.map((el) => `(${el})`).join("")}`);
  }

  const all = result.join("");
  return `(&${all})`;
}


src/helpers/tsString.ts
import { Query, WhereInput } from "../index";
import {
  equal,
  substrings,
  present,
  lessOrEqual,
  greaterOrEqual,
  extensible,
  approxMatch,
  endWith,
  startWith,
} from "./criteria";
import { writeLog } from "fast-node-logger";

function whereClausesToString({
  field,
  criteria,
  action,
  extensibleConfig,
}: WhereInput) {
  writeLog(`whereClausesToString()`, { level: "trace" });
  if (action === "equal") {
    return equal({ field: field as string, criteria });
  }
  if (action === "startWith") {
    return startWith({ field: field as string, criteria });
  }
  if (action === "endWith") {
    return endWith({ field: field as string, criteria });
  }
  if (action === "approxMatch") {
    return approxMatch({ field: field as string, criteria });
  }
  if (action === "greaterOrEqual") {
    return greaterOrEqual({ field: field as string, criteria });
  }
  if (action === "lessOrEqual") {
    return lessOrEqual({ field: field as string, criteria });
  }
  if (action === "present") {
    return present({ field: field as string });
  }
  if (action === "substrings") {
    return substrings({ field: field as string, criteria });
  }
  if (action === "extensible") {
    return extensible({
      field: extensibleConfig?.ignoreField ? undefined : (field as string),
      criteria,
      dn: extensibleConfig?.dn,
      matchingRuleId: extensibleConfig?.matchingRuleId,
    });
  }
}

/** generate ldap representation of query */
export function toString(query: Query): string {
  writeLog(`toString()`, { level: "trace" });
  const result: string[] = [];

  if (query.where) {
    result.push(`(${whereClausesToString(query.where)})`);
  }

  if (query.whereAnd.length > 0) {
    result.push(
      `(&${query.whereAnd
        .map((el) => `(${whereClausesToString(el)})`)
        .join("")})`,
    );
  }

  if (query.whereOr.length > 0) {
    result.push(
      `(|${query.whereOr
        .map((el) => `(${whereClausesToString(el)})`)
        .join("")})`,
    );
  }

  if (query.whereNot.length > 0) {
    result.push(
      `(!${query.whereNot
        .map((el) => `(${whereClausesToString(el)})`)
        .join("")})`,
    );
  }

  if (query.whereRaw.length > 0) {
    result.push(`${query.whereRaw.map((el) => `(${el})`).join("")}`);
  }

  const all = result.join("");
  return `(&${all})`;
}


src/app.ts
import dotenv from "dotenv";
dotenv.config();
import { createLogger, writeLog } from "fast-node-logger";
import { QueryGenerator } from "./index";

export async function main() {
  const logger = await createLogger({
    level: "trace",
  });

  // const qGen = new QueryGenerator({ logger });
  // const { query } = qGen
  //   .select(["cn"])
  //   .where({
  //     field: "userPrincipalName",
  //     action: "substrings",
  //     criteria: "sostad*",
  //   })
  //   .whereAnd({ field: "objectClass", action: "equal", criteria: "user" })
  //   .whereOr({ field: "objectClass", action: "equal", criteria: "person" })
  //   .whereNot({ field: "objectClass", action: "equal", criteria: "computer" })
  //   .whereNot({ field: "objectClass", action: "equal", criteria: "group" });

  // const queryString = query.toString();

  // // validation
  // (function queryValidate(str: string) {
  //   let result = "";
  //   /* Wrap the input in parens if it was not already */
  //   if (str.charAt(0) !== '(') {
  //     result = `(${str})`;
  //   }
  //   const end = result.length;

  //   console.log(`File: app.ts,`, `Line: 34 => `, result)

  // })(queryString)

  const qGen2 = new QueryGenerator({ logger });
  const { query } = qGen2
    .select(["USNIntersite", "aCSPolicyName"])
    .where({ field: "mobile", action: "substrings", criteria: "404*999*" })
    .whereAnd({ field: "memberOf", action: "startWith", criteria: "admin" })
    .whereAnd({ field: "memberOf", action: "endWith", criteria: "office" })
    .whereAnd({ field: "badPwdCount", action: "lessOrEqual", criteria: "2" })
    .whereAnd({ field: "info", action: "approxMatch", criteria: "my-info" })
    .whereOr({ field: "mail", action: "present", criteria: "*@domain.com" })
    .whereOr({
      field: "homePostalAddress",
      action: "substrings",
      criteria: "Georgia",
    })
    .whereNot({
      field: "delivContLength",
      action: "greaterOrEqual",
      criteria: "6",
    })
    .whereNot({
      field: "middleName",
      action: "extensible",
      criteria: "joe",
      extensibleConfig: {
        dn: true,
        ignoreField: true,
        matchingRuleId: "1.2.840.113556.1.4.1941",
      },
    })
    .whereNot({
      field: "userAccountControl",
      action: "extensible",
      criteria: "2",
      extensibleConfig: {
        dn: false,
        ignoreField: false,
        matchingRuleId: "1.2.840.113556.1.4.803",
      },
    })
    .whereRaw("&(cn=3)(dn=*)")
    .whereRaw("phone=*11");

  console.log(`File: app.ts,`, `Line: 22 => `, query.toString());

  // const qGen2 = new QueryGenerator({ logger });
  // const { query: query2 } = qGen2
  //   .select(["USNIntersite", "aCSPolicyName", "accountNameHistory"])
  //   .where({ field: "mobile", action: "substrings", criteria: "404*999*" })
  //   .whereAnd({ field: "memberOf", action: "startWith", criteria: "admin" })
  //   .whereAnd({ field: "memberOf", action: "endWith", criteria: "office" })
  //   .whereAnd({ field: "badPwdCount", action: "lessOrEqual", criteria: "2" })
  //   .whereAnd({ field: "info", action: "approxMatch", criteria: "my-info" })
  //   .whereOr({ field: "mail", action: "present", criteria: "*@domain.com" })
  //   .whereOr({
  //     field: "homePostalAddress",
  //     action: "substrings",
  //     criteria: "Georgia",
  //   })
  //   .whereNot({
  //     field: "delivContLength",
  //     action: "greaterOrEqual",
  //     criteria: "6",
  //   })
  //   .whereNot({
  //     field: "middleName",
  //     action: "extensible",
  //     criteria: "joe",
  //     extensibleConfig: {
  //       dn: true,
  //       ignoreField: true,
  //       matchingRuleId: "1.2.840.113556.1.4.1941",
  //     },
  //   })
  //   .whereRaw("&(cn=3)(dn=*)")
  //   .whereRaw("phone=*11");

  // console.log(`File: app.ts,`, `Line: 22 => `, query2.toString());
}

main().catch((err: Error) => {
  writeLog(err, { level: "error", stdout: true });
});
